File: component-vector/build_icon_index.js
#!/usr/bin/env node
'use strict';

/**
 * 构建图标向量索引，写入 Elasticsearch。
 *
 * 用法：
 *   node build_icon_index.js          # 真实 embedding API
 *   MOCK_EMBED=1 node build_icon_index.js  # mock 随机向量
 */

const fs     = require('fs');
const path   = require('path');
const OpenAI = require('openai');
const { getClient } = require('./es_client');

// ── 环境配置 ─────────────────────────────────────────────────────────────────

const envFile = path.resolve(__dirname, '.env');
if (fs.existsSync(envFile)) {
  fs.readFileSync(envFile, 'utf8').split('\n').forEach(line => {
    const t = line.trim();
    if (!t || t.startsWith('#')) return;
    const idx = t.indexOf('=');
    if (idx < 0) return;
    const k = t.slice(0, idx).trim();
    const v = t.slice(idx + 1).trim();
    if (k && !process.env[k]) process.env[k] = v;
  });
}

const MOCK_MODE          = /^(1|true|yes)$/i.test(process.env.MOCK_EMBED || '');
const EMBEDDING_BASE_URL = process.env.EMBEDDING_BASE_URL || 'https://dashscope.aliyuncs.com/compatible-mode/v1';
const EMBEDDING_MODEL    = process.env.EMBEDDING_MODEL    || 'text-embedding-v3';
const EMBEDDING_DIM      = Number(process.env.EMBEDDING_DIM) || 1024;
const BATCH_SIZE         = 25;

const ICONS_PATH = process.env.ICONS_PATH || path.resolve(__dirname, 'icons.json');
const ES_INDEX   = process.env.ICON_ES_INDEX || 'component_icons';
const REINDEX    = /^(1|true|yes)$/i.test(process.env.REINDEX || '');

// ── Embedding 文本构建 ────────────────────────────────────────────────────────

// 判断是否为自然语言（含中文或空格）
function isNaturalLanguage(str) {
  return /[一-鿿，。、\s]/.test(str);
}

// 将标识符风格的字符串拆成可读词组
// "ic_it_terminology_qa" → "it terminology qa"
// "terminologyQA"        → "terminology QA"
function splitIdentifier(str) {
  return str
    .replace(/^ic_/, '')                        // 去掉 ic_ 前缀
    .replace(/([a-z])([A-Z])/g, '$1 $2')        // camelCase 拆词
    .replace(/_/g, ' ')                          // 下划线转空格
    .trim();
}

function buildIconText(icon) {
  const parts = [icon.name];

  if (icon.englishName) {
    parts.push(
      isNaturalLanguage(icon.englishName)
        ? icon.englishName
        : splitIdentifier(icon.englishName)
    );
  }

  if (icon.description) {
    parts.push(
      isNaturalLanguage(icon.description)
        ? icon.description
        : splitIdentifier(icon.description)
    );
  }

  return parts.filter(Boolean).join(' ');
}

// ── Embedding ─────────────────────────────────────────────────────────────────

const client = new OpenAI({
  apiKey:  process.env.DASHSCOPE_API_KEY,
  baseURL: EMBEDDING_BASE_URL,
  timeout: 60_000,
});

function mockEmbedding() {
  const v = Array.from({ length: EMBEDDING_DIM }, () => Math.random() * 2 - 1);
  const norm = Math.sqrt(v.reduce((s, x) => s + x * x, 0));
  return v.map(x => x / norm);
}

async function embedBatch(texts) {
  if (MOCK_MODE) return texts.map(() => mockEmbedding());
  const resp = await client.embeddings.create({
    model:           EMBEDDING_MODEL,
    input:           texts,
    dimensions:      EMBEDDING_DIM,
    encoding_format: 'float',
  });
  return resp.data.map(d => d.embedding);
}

// ── ES 操作 ───────────────────────────────────────────────────────────────────

const ES_MAPPING = {
  mappings: {
    properties: {
      text:        { type: 'text' },
      embedding:   { type: 'dense_vector', dims: EMBEDDING_DIM, index: true, similarity: 'cosine' },
      icon_id:      { type: 'keyword' },
      name:         { type: 'keyword' },
      english_name: { type: 'keyword' },
      description:  { type: 'text' },
    },
  },
};

async function ensureIndex(es) {
  const exists = await es.indices.exists({ index: ES_INDEX });
  if (exists) {
    if (REINDEX) {
      console.log(`索引 ${ES_INDEX} 已存在，追加模式`);
      return;
    }
    console.log(`删除旧索引 ${ES_INDEX} ...`);
    await es.indices.delete({ index: ES_INDEX });
  }
  console.log(`创建索引 ${ES_INDEX} ...`);
  await es.indices.create({ index: ES_INDEX, ...ES_MAPPING });
}

async function bulkIndex(es, docs) {
  const body = docs.flatMap(doc => [
    { index: { _index: ES_INDEX } },
    doc,
  ]);
  const resp = await es.bulk({ body, refresh: false });
  if (resp.errors) {
    const failed = resp.items.filter(i => i.index?.error);
    console.warn(`  ⚠ 本批有 ${failed.length} 条写入失败`);
  }
}

// ── main ─────────────────────────────────────────────────────────────────────

async function main() {
  console.log(MOCK_MODE ? '[MOCK 模式] 使用随机向量' : `[真实模式] ${EMBEDDING_BASE_URL} / ${EMBEDDING_MODEL}`);

  const es = getClient();
  await es.ping().catch(() => { throw new Error(`无法连接 ES：${process.env.ES_URL || 'http://localhost:9200'}`); });
  console.log('ES 连接正常');

  const icons = JSON.parse(fs.readFileSync(ICONS_PATH, 'utf8'));
  console.log(`共 ${icons.length} 个图标`);

  // 预览
  console.log('\n--- embedding 文本预览 ---');
  icons.slice(0, 3).forEach((icon, i) => {
    console.log(`[${i}] ${buildIconText(icon).slice(0, 100)}`);
  });
  console.log('...\n');

  await ensureIndex(es);

  const ES_BULK      = 200;
  const embedBuf     = [];
  const totalBatches = Math.ceil(icons.length / BATCH_SIZE);

  for (let i = 0; i < icons.length; i += BATCH_SIZE) {
    const batch = icons.slice(i, i + BATCH_SIZE);
    process.stdout.write(`Embed 批次 ${Math.floor(i / BATCH_SIZE) + 1}/${totalBatches} ... `);

    const texts = batch.map(icon => buildIconText(icon));
    const vecs  = await embedBatch(texts);
    console.log('done');

    for (let j = 0; j < batch.length; j++) {
      const icon = batch[j];
      embedBuf.push({
        text:         buildIconText(icon),
        embedding:    vecs[j],
        icon_id:      icon.id,
        name:         icon.name,
        english_name: icon.englishName || '',
        description:  icon.description || '',
      });
    }

    if (embedBuf.length >= ES_BULK || i + BATCH_SIZE >= icons.length) {
      await bulkIndex(es, embedBuf.splice(0));
      process.stdout.write(`  → ES 写入至 ${Math.min(i + BATCH_SIZE, icons.length)} 条\n`);
    }

    if (!MOCK_MODE && i + BATCH_SIZE < icons.length) {
      await new Promise(r => setTimeout(r, 150));
    }
  }

  await es.indices.refresh({ index: ES_INDEX });
  const count = await es.count({ index: ES_INDEX });
  console.log(`\n完成！ES 索引 ${ES_INDEX} 共 ${count.count} 条文档`);
}

main().catch(err => { console.error('Fatal:', err.message); process.exit(1); });

File: component-vector/build_icon_index.py
#!/usr/bin/env python3
"""
构建图标向量索引，写入 Elasticsearch。

用法：
  python build_icon_index.py
  MOCK_EMBED=1 python build_icon_index.py
"""

import json
import re
import sys
from pathlib import Path
from typing import Dict, List

from config import (
    MOCK_MODE, EMBEDDING_BASE_URL, EMBEDDING_MODEL,
    ICONS_PATH, ICON_ES_INDEX, REINDEX, EMBEDDING_DIM,
)
from es_client import get_client
from embed_client import embed_many

# ── Embedding 文本构建 ────────────────────────────────────────────────────────

def _is_natural_language(s: str) -> bool:
    return bool(re.search(r'[一-鿿，。、\s]', s))


def _split_identifier(s: str) -> str:
    s = re.sub(r'^ic_', '', s)
    s = re.sub(r'([a-z])([A-Z])', r'\1 \2', s)
    return s.replace('_', ' ').strip()


def build_icon_text(icon: dict) -> str:
    parts = [icon.get('name', '')]

    en = icon.get('englishName', '')
    if en:
        parts.append(en if _is_natural_language(en) else _split_identifier(en))

    desc = icon.get('description', '')
    if desc:
        parts.append(desc if _is_natural_language(desc) else _split_identifier(desc))

    return ' '.join(filter(None, parts))


# ── ES 操作 ───────────────────────────────────────────────────────────────────

ES_MAPPING = {
    'properties': {
        'text':         {'type': 'text'},
        'embedding':    {'type': 'dense_vector', 'dims': EMBEDDING_DIM, 'index': True, 'similarity': 'cosine'},
        'icon_id':      {'type': 'keyword'},
        'name':         {'type': 'keyword'},
        'english_name': {'type': 'keyword'},
        'description':  {'type': 'text'},
    }
}


def ensure_index(es):
    exists = es.indices.exists(index=ICON_ES_INDEX)
    if exists:
        if REINDEX:
            print(f'索引 {ICON_ES_INDEX} 已存在，追加模式（REINDEX=1）')
            return
        print(f'删除旧索引 {ICON_ES_INDEX} ...')
        es.indices.delete(index=ICON_ES_INDEX)
    print(f'创建索引 {ICON_ES_INDEX} ...')
    es.indices.create(index=ICON_ES_INDEX, mappings=ES_MAPPING)


def bulk_index(es, docs: List[Dict]):
    operations = []
    for doc in docs:
        operations.append({'index': {'_index': ICON_ES_INDEX}})
        operations.append(doc)
    resp = es.bulk(operations=operations)
    if resp.get('errors'):
        failed = [item for item in resp['items'] if item.get('index', {}).get('error')]
        print(f'  ⚠ 本批有 {len(failed)} 条写入失败')


# ── main ──────────────────────────────────────────────────────────────────────

def main():
    print('[MOCK 模式] 使用随机向量' if MOCK_MODE else f'[真实模式] {EMBEDDING_BASE_URL} / {EMBEDDING_MODEL}')

    es = get_client()
    if not es.ping():
        print('无法连接 ES，请检查 ES_URL 配置', file=sys.stderr)
        sys.exit(1)
    print('ES 连接正常')

    icons = json.loads(Path(ICONS_PATH).read_text())
    print(f'共 {len(icons)} 个图标')

    print('\n--- embedding 文本预览 ---')
    for i, icon in enumerate(icons[:3]):
        print(f'[{i}] {build_icon_text(icon)[:100]}')
    print('...\n')

    ensure_index(es)

    BATCH_SIZE = 25
    ES_BULK    = 200
    embed_buf  = []
    total      = len(icons)

    for i in range(0, total, BATCH_SIZE):
        batch = icons[i:i + BATCH_SIZE]
        batch_no = i // BATCH_SIZE + 1
        total_batches = (total + BATCH_SIZE - 1) // BATCH_SIZE
        print(f'Embed 批次 {batch_no}/{total_batches} ...', end=' ', flush=True)

        texts = [build_icon_text(icon) for icon in batch]
        vecs  = embed_many(texts)
        print('done')

        for icon, text, vec in zip(batch, texts, vecs):
            embed_buf.append({
                'text':         text,
                'embedding':    vec,
                'icon_id':      icon.get('id', ''),
                'name':         icon.get('name', ''),
                'english_name': icon.get('englishName', ''),
                'description':  icon.get('description', ''),
            })

        if len(embed_buf) >= ES_BULK or i + BATCH_SIZE >= total:
            bulk_index(es, embed_buf)
            embed_buf.clear()
            print(f'  → ES 写入至 {min(i + BATCH_SIZE, total)} 条')

    es.indices.refresh(index=ICON_ES_INDEX)
    count = es.count(index=ICON_ES_INDEX)['count']
    print(f'\n完成！ES 索引 {ICON_ES_INDEX} 共 {count} 条文档')


if __name__ == '__main__':
    main()
